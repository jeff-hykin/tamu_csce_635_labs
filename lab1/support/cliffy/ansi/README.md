# Ansi

<p align="center" class="badges-container">
  <a href="https://github.com/c4spar/deno-cliffy/releases">
    <img alt="Version" src="https://img.shields.io/github/v/release/c4spar/deno-cliffy?logo=github&color=blue" />
  </a>
  <a href="https://github.com/c4spar/deno-cliffy/actions/workflows/test.yml">
    <img alt="Build status" src="https://github.com/c4spar/deno-cliffy/workflows/Test/badge.svg?branch=main" />
  </a>
  <a href="https://codecov.io/gh/c4spar/deno-cliffy">
    <img src="https://codecov.io/gh/c4spar/deno-cliffy/branch/main/graph/badge.svg"/>
  </a>
  <a href="https://github.com/c4spar/deno-cliffy/labels/module%3Aansi-escape">
    <img alt="issues" src="https://img.shields.io/github/issues/c4spar/deno-cliffy/module:ansi-escape?label=issues&logo=github&color=yellow">
  </a>
  <a href="https://deno.land/">
    <img alt="Deno version" src="https://img.shields.io/badge/deno-^1.4.0-blue?logo=deno" />
  </a>
  <a href="https://doc.deno.land/https/deno.land/x/cliffy/ansi/mod.ts">
    <img alt="doc" src="https://img.shields.io/badge/deno-doc-yellow?logo=deno" />
  </a>
  <a href="https://discord.gg/ghFYyP53jb">
    <img alt="Discord" src="https://img.shields.io/badge/join-chat-blue?logo=discord&logoColor=white" />
  </a>
  <a href="../LICENSE">
    <img alt="License" src="https://img.shields.io/github/license/c4spar/deno-cliffy?logo=github" />
  </a>
  <br>
  <a href="https://deno.land/x/cliffy">
    <img alt="Discord" src="https://img.shields.io/badge/Published on deno.land-blue?logo=deno&logoColor=959DA6&color=272727" />
  </a>
</p>

The documentation has been moved to https://cliffy.io/docs/ansi.
