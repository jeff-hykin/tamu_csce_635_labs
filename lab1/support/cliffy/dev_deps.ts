/* std */
export {
  assert,
  assertEquals,
  assertInstanceOf,
  assertRejects,
  assertStrictEquals,
  assertThrows,
} from "https://deno.land/std@0.196.0/testing/asserts.ts";
export {
  assertSpyCall,
  assertSpyCalls,
  spy,
} from "https://deno.land/std@0.196.0/testing/mock.ts";
export { assertSnapshot } from "https://deno.land/std@0.196.0/testing/snapshot.ts";
export { describe, it } from "https://deno.land/std@0.196.0/testing/bdd.ts";
export {
  assertType,
  type IsExact,
} from "https://deno.land/std@0.196.0/testing/types.ts";
export {
  bold,
  red,
  stripColor,
} from "https://deno.land/std@0.196.0/fmt/colors.ts";
export { dirname } from "https://deno.land/std@0.196.0/path/mod.ts";
export { expandGlob } from "https://deno.land/std@0.196.0/fs/expand_glob.ts";
export type { WalkEntry } from "https://deno.land/std@0.196.0/fs/walk.ts";
export { copy } from "https://deno.land/std@0.196.0/streams/copy.ts";
export { format } from "https://deno.land/std@0.196.0/datetime/format.ts";
export { gt } from "https://deno.land/std@0.196.0/semver/gt.ts";
export { lt } from "https://deno.land/std@0.196.0/semver/lt.ts";

/* 3rd party */
export { default as sinon } from "https://cdn.skypack.dev/sinon@v13.0.2?dts";
