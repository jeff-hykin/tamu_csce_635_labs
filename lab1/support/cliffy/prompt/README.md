# Prompt

The documentation has been moved to https://cliffy.io/docs/prompt.
