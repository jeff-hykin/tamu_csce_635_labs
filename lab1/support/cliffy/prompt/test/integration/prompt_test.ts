import { ansi } from "../../../ansi/ansi.ts";
import { assertEquals, assertType, IsExact } from "../../../dev_deps.ts";
import { Checkbox, CheckboxOptions } from "../../checkbox.ts";
import { Confirm } from "../../confirm.ts";
import { Input } from "../../input.ts";
import { Number } from "../../number.ts";
import { prompt, PromptMiddleware, PromptOptions } from "../../prompt.ts";
import { Select } from "../../select.ts";
import { Toggle } from "../../toggle.ts";
import { snapshotTest } from "../../../testing/snapshot.ts";

await snapshotTest({
  name: "prompt method",
  meta: import.meta,
  osSuffix: ["windows"],
  steps: {
    "should check an option": {
      stdin: ansi
        // input
        .text("f")
        .text("o")
        .text("o")
        .text("\n")
        // number
        .text("4")
        .text("3")
        .text("\n")
        // confirm
        .text("y")
        .text("\n")
        // toggle
        .text("y")
        .text("n")
        .text("y")
        .text("\n")
        // checkbox
        .cursorDown
        .text(" ")
        .cursorDown
        .text(" ")
        .text("\n")
        .text("\n")
        // select
        .cursorDown
        .text("\n")
        .toArray(),
    },
  },
  async fn() {
    const checkboxOptions: PromptOptions<
      "checkbox",
      typeof Checkbox,
      { input?: string }
    > = {
      name: "checkbox",
      type: Checkbox,
      message: "Select an option",
      options: [
        { name: "Foo", value: "foo" },
        { name: "Bar", value: "bar" },
        { name: "Baz", value: "baz" },
      ],
      before({ input }, next) {
        console.log("[checkbox:before] input is:", input);
        return next();
      },
    };

    const result = await prompt([
      {
        name: "input",
        type: Input,
        message: "Enter some text",
        default: "default value",
      },
      {
        name: "number",
        type: Number,
        message: "Enter a number",
      },
      {
        name: "confirm",
        type: Confirm,
        message: "Please confirm",
      },
      {
        name: "toggle",
        type: Toggle,
        message: "Please toggle",
        hint: "some hint",
        default: false,
      },
      checkboxOptions,
      {
        name: "select",
        type: Select<number>,
        message: "Select a value",
        options: [1, 2, 3],
      },
    ], {
      reader: Deno.stdin,
      writer: Deno.stderr,
    });

    assertType<
      IsExact<typeof result, {
        select?: number | undefined;
        checkbox?: string[] | undefined;
        input?: string | undefined;
        number?: number | undefined;
        confirm?: boolean | undefined;
        toggle?: boolean | undefined;
      }>
    >(true);

    assertEquals(result, {
      input: "foo",
      number: 43,
      confirm: true,
      toggle: true,
      checkbox: ["bar", "baz"],
      select: 2,
    });

    assertType<
      IsExact<
        typeof checkboxOptions,
        {
          name: "checkbox";
          type: typeof Checkbox;
          before?: PromptMiddleware<
            { checkbox?: Array<string>; input?: string }
          >;
          after?: PromptMiddleware<
            { checkbox?: Array<string>; input?: string }
          >;
        } & CheckboxOptions<string>
      >
    >(true);

    console.log(result);
  },
});
