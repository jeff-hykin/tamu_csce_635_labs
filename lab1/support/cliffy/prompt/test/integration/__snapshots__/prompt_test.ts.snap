export const snapshot = {};

snapshot[`prompt method > should check an option 1`] = `
stdout:
'[checkbox:before] input is: foo
{
  input: "foo",
  number: 43,
  confirm: true,
  toggle: true,
  checkbox: [ "bar", "baz" ],
  select: 2
}
'
stderr:
"? Enter some text (default value) › \\x1b[37G\\x1b[G\\x1b[0J
? Enter some text (default value) › f\\x1b[38G\\x1b[G\\x1b[0J
? Enter some text (default value) › fo\\x1b[39G\\x1b[G\\x1b[0J
? Enter some text (default value) › foo\\x1b[40G\\x1b[G\\x1b[0J
? Enter some text (default value) › foo
\\x1b[?25h\\x1b[?25h\\x1b[?25h? Enter a number › \\x1b[20G\\x1b[G\\x1b[0J
? Enter a number › 4\\x1b[21G\\x1b[G\\x1b[0J
? Enter a number › 43\\x1b[22G\\x1b[G\\x1b[0J
? Enter a number › 43
\\x1b[?25h\\x1b[?25h\\x1b[?25h? Please confirm (y/n) › \\x1b[26G\\x1b[G\\x1b[0J
? Please confirm (y/n) › yes\\x1b[27G\\x1b[G\\x1b[0J
? Please confirm (y/n) › Yes
\\x1b[?25h\\x1b[?25h\\x1b[?25h? Please toggle (No) › No / Yes
❯ some hint\\x1b[1A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Please toggle (No) › No / Yes
❯ some hint\\x1b[1A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Please toggle (No) › No / Yes
❯ some hint\\x1b[1A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Please toggle (No) › No / Yes
❯ some hint\\x1b[1A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Please toggle (No) › Yes
\\x1b[?25h\\x1b[?25h\\x1b[?25h? Select an option
❯ ✘ Foo
  ✘ Bar
  ✘ Baz\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  ✘ Foo
❯ ✘ Bar
  ✘ Baz\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  ✘ Foo
❯ ✔ Bar
  ✘ Baz\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  ✘ Foo
  ✔ Bar
❯ ✘ Baz\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  ✘ Foo
  ✔ Bar
❯ ✔ Baz\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  ✘ Foo
  ✔ Bar
❯ ✔ Baz
Press ↵ again to submit. To check or uncheck the selected option press space.\\x1b[4A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option › Bar, Baz
\\x1b[?25h\\x1b[?25h\\x1b[?25h? Select a value
❯ 1
  2
  3\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select a value
  1
❯ 2
  3\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select a value › 2
\\x1b[?25h\\x1b[?25h\\x1b[?25h"
`;
